evm2wast.c

```diff
@@ -62,11 +77,18 @@ int evm2wasm(char *evm_code, size_t len, char *wast_code, size_t wast_size)
 		}
 		wast_chunk = gadgets[(int)evm_code[i]];
 		wast_code = realloc(wast_code, strlen(wast_code)+strlen(wast_chunk) + 3);
+		if(wast_code == 0) goto err;
 		wast_code = strcat(wast_code, wast_chunk);
 	}
 	char *with_segments = assemble_segments(jump_segments, 4);
 	wast_code = realloc(wast_code, strlen(wast_code) + strlen(with_segments));
+	if(wast_code == 0) goto err;
```

Comment:
Note: realloc doesn't free the original memory space on failure.

Reply:
only if the first param and the return value is different
`b = realloc(a, size)` will leave a untouched and set b to 0.

Reply:
`realloc` has no understanding what the lvalue is.

See the man pages:
> For realloc(), the input pointer is still valid if reallocation failed.  For reallocf(), the input pointer will have been freed if reallocation failed.

Reply:
```
The  realloc()  function  returns a pointer to the newly allocated memory,
       which is suitably aligned for any built-in type and may be different  from
       ptr,  or NULL if the request fails
....
 If  realloc()
       fails, the original block is left untouched; it is not freed or moved.
```
implementation dependent but as the README.md states we go with GNU

Reply:
I do not follow.

For one, doesn't matter what the README says, you cannot claim "go with glibc" since it will definitely not use glibc when compiling as a contract.

Second, I think it is clear what the manual is saying. The input pointer passed to `realloc` is not freed upon failure.

Reply:
emcc is using glibc afaik.
Your comment applies to my check `wasm_code` being 0, which will happen on `realloc` failure for it will return with a location either way.
If you refer to a potential `strlen(wasm_code)` leak here, that could happen and we can mitigate with a macro or inline